pykafka/producer.py

```diff
@@ -317,30 +344,12 @@ def _get_partition_msgs(partition_id, req):
                 time.sleep(self._retry_backoff_ms / 1000)
                 self._prepare_request(to_retry, owned_broker, attempt)
             else:
-                raise ProduceFailureError('Unable to produce messages. See log for details.')
-
-    def _update_leaders(self):
-        """Ensure each message in each queue is in the queue owned by its
-            partition's leader
-
-        This function empties all broker queues, maps their messages to the
-        current leaders for their partitions, and enqueues the messages in
-        the appropriate queues.
-        """
-        # empty queues and figure out updated partition leaders
-        new_queue_contents = defaultdict(list)
-        for owned_broker in itervalues(self._owned_brokers):
-            owned_broker.lock.acquire()
-            current_queue_contents = owned_broker.flush(0, release_pending=True)
-            for kv, partition_id in current_queue_contents:
-                partition_leader = self._topic.partitions[partition_id].leader
-                new_queue_contents[partition_leader.id].append((kv, partition_id))
-        # retain locks for all brokers between these two steps
-        for owned_broker in itervalues(self._owned_brokers):
-            owned_broker.enqueue(new_queue_contents[owned_broker.broker.id],
-                                 self._block_on_queue_full)
-            owned_broker.resolve_event_state()
-            owned_broker.lock.release()
+                log.error("Failed to produce messages to broker %s:%s after %s attempts. "
+                          "Re-enqueuing %s messages.", owned_broker.broker.host,
+                          owned_broker.broker.port, attempt, len(to_retry))
+                owned_broker.increment_messages_pending(-1 * len(to_retry))
+                for tup in to_retry:
+                    self._produce(tup)
```

Comment:
Ah right - I didn't sufficiently realise that the `ProduceFailureError` would crash the worker thread.

Sorry, yes that was silly - so I should have argued that instead of `raise ProduceFailureError` it should simply say `pass` and drop the message. That wouldn't crash the worker, and it's consistent with what the user asks, ie give up after `max_retries`.

(And then when we start returning futures from `produce()`, we can then push a `ProduceFailureError` onto the future instead, but that's a separate issue.)